File: tests/workflow/producer/rust/src/producer/implementation/handlers/structa.rs

use super::{
    broadcast, identification, pack, producer::Control, protocol, responses, Context, HandlerError,
    ProducerError, scope::Scope,
};
use clibri::server;
use uuid::Uuid;

pub async fn process<E: server::Error, C: server::Control<E>>(
    identification: &identification::Identification,
    filter: &identification::Filter<'_>,
    context: &mut Context,
    request: &protocol::StructA,
    sequence: u32,
    control: &Control<E, C>,
) -> Result<(), HandlerError> {
    let mut scope: Scope<'_, E, C> = Scope::new(context, control, identification, filter);
    let uuid = identification.uuid();
    let mut broadcasting: Vec<(Vec<Uuid>, Vec<u8>)> = vec![];
    let buffer =
        match responses::structa::response(request, &mut scope).await {
            Ok(conclusion) => match conclusion {
                responses::structa::Response::CaseB((
                    mut response,
                	mut broadcast_structd,
                	mut broadcast_structf,
                )) => {
                    broadcasting.push((
                        broadcast_structd.0,
                        pack(&0, &uuid, &mut broadcast_structd.1)?,
                    ));
                    broadcasting.push((
                        broadcast_structf.0,
                        pack(&0, &uuid, &mut broadcast_structf.1)?,
                    ));
                    pack(&sequence, &uuid, &mut response)?
                },
                responses::structa::Response::CaseC(mut response) => {
                    pack(&sequence, &uuid, &mut response)?
                },
                responses::structa::Response::CaseD((
                    mut response,
                	mut broadcast_structj,
                )) => {
                    broadcasting.push((
                        broadcast_structj.0,
                        pack(&0, &uuid, &mut broadcast_structj.1)?,
                    ));
                    pack(&sequence, &uuid, &mut response)?
                },
            },
            Err(mut error) => pack(&sequence, &uuid, &mut error)?,
        };
    control
        .send(buffer, Some(uuid))
        .await
        .map_err(|e: ProducerError<E>| HandlerError::Processing(e.to_string()))?;
    for msg in broadcasting.iter_mut() {
        broadcast::<E, C>(msg, control).await?;
    }
    scope.call().await;
    Ok(())
}    

File: tests/workflow/producer/typescript/src/producer/implementation/responses/structa.ts
import {
    Producer,
    Identification,
    Filter,
    broadcastAll,
    Context,
    Protocol,
} from "./index";
import { response } from "../../responses/structa";
import { Scope } from "../scope";

export class Response {    
    static REQUIRED_CASEB = [    
        Protocol.StructD,
        Protocol.StructF,
    ];
    static REQUIRED_CASED = [    
        Protocol.StructJ,
    ];
    private _response!: Protocol.StructB | Protocol.StructC | Protocol.StructD | Protocol.StructE;
    private _broadcasts: Array<[string[], Protocol.Convertor<any>]> = [];

    constructor(
        res: Protocol.StructB | Protocol.StructC | Protocol.StructD | Protocol.StructE
    ) {
        this._response = res;
    }

    public broadcast(uuids: string[]): {        
        StructD(msg: Protocol.StructD): Response;
        StructF(msg: Protocol.StructF): Response;
        StructJ(msg: Protocol.StructJ): Response;
    } {
        const self = this;
        return {            
            StructD(msg: Protocol.StructD): Response {
                if (
                    self._response.getId() !==
                    Protocol.StructB.getId()
                ) {
                    throw new Error(
                        `Message "Protocol.StructD" can be used only with "Protocol.StructB"`
                    );
                }
                if (
                    self._broadcasts.find(
                        (b) =>
                            b[1].getId() ===
                            Protocol.StructD.getId()
                    ) !== undefined
                ) {
                    throw new Error(
                        `Broadcast Protocol.StructD already has been defined.`
                    );
                }
                self._broadcasts.push([uuids, msg]);
                return self;
            },
            StructF(msg: Protocol.StructF): Response {
                if (
                    self._response.getId() !==
                    Protocol.StructB.getId()
                ) {
                    throw new Error(
                        `Message "Protocol.StructF" can be used only with "Protocol.StructB"`
                    );
                }
                if (
                    self._broadcasts.find(
                        (b) =>
                            b[1].getId() ===
                            Protocol.StructF.getId()
                    ) !== undefined
                ) {
                    throw new Error(
                        `Broadcast Protocol.StructF already has been defined.`
                    );
                }
                self._broadcasts.push([uuids, msg]);
                return self;
            },
            StructJ(msg: Protocol.StructJ): Response {
                if (
                    self._response.getId() !==
                    Protocol.StructD.getId()
                ) {
                    throw new Error(
                        `Message "Protocol.StructJ" can be used only with "Protocol.StructD"`
                    );
                }
                if (
                    self._broadcasts.find(
                        (b) =>
                            b[1].getId() ===
                            Protocol.StructJ.getId()
                    ) !== undefined
                ) {
                    throw new Error(
                        `Broadcast Protocol.StructJ already has been defined.`
                    );
                }
                self._broadcasts.push([uuids, msg]);
                return self;
            },
        };
    }

    public error(): Error | undefined {
        let error: Error | undefined;        
        if (
            error === undefined &&
            this._response.getId() ===
            Protocol.StructB.getId()
        ) {
            Response.REQUIRED_CASEB.forEach((ref) => {
                if (error !== undefined) {
                    return;
                }
                if (
                    this._broadcasts.find((msg) => {
                        return msg[1].getId() === ref.getId();
                    }) === undefined
                ) {
                    error = new Error(
                        `Broadcast ${ref.getSignature()}/${ref.getId()} is required for ${this._response.getSignature()}/${this._response.getId()}, but hasn't been found`
                    );
                }
            });
        }
        if (
            error === undefined &&
            this._response.getId() ===
            Protocol.StructD.getId()
        ) {
            Response.REQUIRED_CASED.forEach((ref) => {
                if (error !== undefined) {
                    return;
                }
                if (
                    this._broadcasts.find((msg) => {
                        return msg[1].getId() === ref.getId();
                    }) === undefined
                ) {
                    error = new Error(
                        `Broadcast ${ref.getSignature()}/${ref.getId()} is required for ${this._response.getSignature()}/${this._response.getId()}, but hasn't been found`
                    );
                }
            });
        }
        return error;
    }

    public pack(sequence: number, uuid: string): ArrayBufferLike {
        return this._response.pack(sequence, uuid);
    }

    public broadcasts(): Array<[string[], Protocol.Convertor<any>]> {
        return this._broadcasts;
    }
}

export function handler(
    request: Protocol.StructA,
    consumer: Identification,
    filter: Filter,
    context: Context,
    producer: Producer,
    sequence: number
): Promise<void> {
	return new Promise((resolve, reject) => {
        const scope = new Scope(consumer, filter, context, producer);
		response(request, scope)
			.then((res) => {
				const error: Error | undefined = res.error();
				if (error instanceof Error) {
					return reject(error);
				}
				producer
					.send(consumer.uuid(), res.pack(sequence, consumer.uuid()))
					.then(() => {
						broadcastAll(producer, res.broadcasts()).then(() => {
                            scope.call();
                            resolve();
                        }).catch(reject);
					}).catch(reject);
			})
			.catch(reject);
	});
}